Storage driver for Google drive

manoj535 opened the pull request:
Pull request for the issue : #1875

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "master" git@github.com:manoj535/distribution.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

RichardScothern commented:
hi @manoj535 .  Thanks for the contribution.  Out of curiosity, what is your use case for storing registry content in google drive?

bharaththiruveedula commented:
Hi @RichardScothern, this PR mostly came because of following requirements.
1)Need a repository which is hosted on cloud for a small team or an individual.
2)Easy to setup(I don't need any setup like OpenStack)
3)And mostly it must be free.

For these type of requirements, people require this type of Cloud storage systems.

dmcgowan commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,488 @@
+}
+
+// New constructs a new Driver with a given rootDirectory
+func New(params DriverParameters) *Driver {
```
This should just return an error rather than call `log.Fatalf`

dmcgowan commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,488 @@
+
+	params := &DriverParameters{
+		KeyFile:       keyFile,
+		RootDirectory: rootDirectory,
```
`RootDirectory` should have a default value set as a const, if no default is appropriate an error should be raised.

dmcgowan commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,488 @@
+// URLFor returns a URL which may be used to retrieve the content stored at the given path.
+// May return an UnsupportedMethodErr in certain StorageDriver implementations.
+func (d *driver) URLFor(ctx context.Context, path string, options map[string]interface{}) (string, error) {
+	return "", storagedriver.ErrUnsupportedMethod{}
```
Is this possible with Google Drive or would the underlying file first need to be shared?

manoj535 replied:
It is not possible with Google Drive api .

dmcgowan commented on Godeps/Godeps.json:
```diff
@@ -198,6 +198,11 @@
 			"Rev": "535138d7bcd717d6531c701ef5933d98b1866257"
 		},
 		{
+			"ImportPath": "github.com/go-ini/ini",
```
This block should be removed. It probably got pulled in by godeps automatically but it is already vendored correctly. We might move away from godeps to something that handles this case better. For now just manually remove it.

manoj535 replied:
@dmcgowan We have removed "github.com/go-ini/ini" from Godeps.json . But ci/circleci build is failing stating "package github.com/go-ini/ini not found"

dmcgowan commented:
This is a good start. I will take a deeper look when I have time to test out the driver myself. Also consider adding a package comment which instructions on the easiest way to set it up. Assume the reader already knows how to access the Google api console. That will make reviewing easier and give a better start on documenting the driver. Also highlight the difference with the GCS driver (note camlistore also has separate backends for Drive and GCS if you find some inspiration there).

More generally on the design, I can see this as being useful for developers and more casual users. I would be really interested to see this be able to integrate with the OAuth work being done so each user could use their own Google Drive storage. I am not expecting any related work in this PR but consider that as a possibility in the way the driver is authorized and setup. Getting that actually working would involve changes higher up the stack to pass in the token.

Thank you for this significant contribution, we really appreciate following up proposals with working code. Going forward is supporting bugs and issues something the two of you would be able to support? We rely on driver contributors to continue to help in supporting their drivers after merge and take that into considering before hitting the merge button.

manoj535 commented:
@dmcgowan We have addressed the comments that you have raised. Please check.
May I know , what kind of differences between gdrive and gcs you are expecting, is it from installation point of view ?

Yes, We are interested in supporting this driver , if there are any issues.

dmcgowan commented on registry/storage/driver/gdrive/gdrive_test.go:
```diff
@@ -0,0 +1,135 @@
+
+func init() {
+
+	keyFile := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS_FILE")
```
It would be advantageous to be able to have the contents of the file be in an environment variable and placed into a tempfile for the tests. The reasoning is that it will be easier for running in a CI environment if it does not need to rely on a file existing.

bharaththiruveedula replied:
Do we need to add a script file in contrib folder that exports environment variable ?

dmcgowan commented on registry/storage/driver/gdrive/gdrive_test.go:
```diff
@@ -0,0 +1,135 @@
+package gdrive
```
Can we add a build tag here `// +build include_gdrive`

bharaththiruveedula replied:
We have added the build tag, but circleci is failing , complaining that 
"no buildable Go source files in  ...."

dmcgowan commented:
Would you mind splitting the commit into a vendor commit and change commit, makes it easier to review on github.

dmcgowan commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,497 @@
+// rootdirectory : Folder name in google drive to store all registry files.
+//
+//
+
```
Should not be a black space here, separate package description

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "master" git@github.com:manoj535/distribution.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

bharaththiruveedula commented:
@dmcgowan We have addressed the comments mentioned, pls check. But the circleci is failing after adding the build tag

stevvooe commented:
How is this different from the gcs driver? Anyone from google willing to support this or coalesce this with the gcs driver?

bharaththiruveedula commented:
Hi @stevvooe as some of the users who wants to store their docker images in easy to setup platform and inexpensive. And this uses google drive API. We have contacted with google for this. Let me why do you  think  this has to be coalesced with gcs driver

stevvooe commented:
@bharaththiruveedula Are you going to be able support this driver? We can't really take on this patch without having someone to support it. I also don't really understand the difference between this and the gcs driver. What is the point of running a production registry against a Google Drive folder?

bharaththiruveedula commented:
@stevvooe 
1)Yes, I would like to support this driver in future too.
2)As @dmcgowan commented in [](https://github.com/docker/distribution/pull/1876#issuecomment-236751378) this is being  useful for developers and more casual users and not specifically for production registry.

stevvooe commented:
@bharaththiruveedula Here is an example of a well-formed proposal: https://github.com/docker/distribution/issues/2041.

I really don't understand the use case and I am not sure if we want to take on the extra technical debt.

aaomidi commented:
This would be a great addition to docker.

mdwheele commented:
Are there any plans to merge this to the project? I just read through and understand the hesitance to accept with regards to support / technical debt, but it looks like @bharaththiruveedula has said repeatedly that there is intent to provide resources for support.

stevvooe commented:
@mdwheele I still haven't received a great answer to what the use case is. Does this run on end user machines? This could be that I misunderstand the differences between google drive and google cloud storage from a product perspective.

Effectively, this PR needs to include some documentation that provides users help with making a choice between the configurations.

@mattmoor Do you have someone who could weigh in on the benefits of drive vs gcs for registry storage?

saschaarthur commented:
would love this feature..

pmjohann commented:
@stevvooe A proper use case is that many people (students for example) are offered Drive storage cheaper than GCS as far as I know. Hence they could benefit from this for example, and I think it would be a good cause to support students by all means, including offering them cheaper storage when learning / using Docker! What do you think?
Keep up the good work guys!

techknowlogick commented:
@stevvooe @pmjohann This is exactly my use case. I'm a student and my school gives me Google Drive for use for my school work. As it is for school I can't publish anything publicly until after the term is up, which means that I can't just use Docker Hub. My school also gives me computation resources so I can run the registry software, but not nearly enough space for the images I would need to store. I already pay for Google Drive through tuition and other associated fees, whereas Google Cloud Storage would be an additional cost.

stevvooe commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,497 @@
+
+func fromParametersImpl(parameters map[string]interface{}) (*DriverParameters, error) {
+
+	keyFile, ok := parameters["keyFile"]
```
Don't use camel case: we only use flat in configuration for registry: `keyfile`.

stevvooe commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,497 @@
+	return New(*params)
+}
+
+func fromParametersImpl(parameters map[string]interface{}) (*DriverParameters, error) {
```
No need to return a pointer here.

stevvooe replied:
Also, why the wrapper func?

stevvooe commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,497 @@
+		}, nil
+	}
+
+	r, _ := d.driveService.Files.List().Q(`"` + d.rootID + `" in parents`).Do()
```
Don't ignore the error here.

stevvooe commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,497 @@
+// ModTime returns the modification time for the file. For backends that
+// don't have a modification time, the creation time should be returned.
+func (fi fileInfo) ModTime() time.Time {
+	t, _ := time.Parse(time.RFC3339, fi.driveFileInfo.modifiedTime)
```
Is this fixed for `time.RFC3339` or is `time.RFC3339Nanos` supported?

stevvooe commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,497 @@
+
+func newFileWriter(bytes []byte, path string, size int, d *driver) *fileWriter {
+	return &fileWriter{
+		bytes:  bytes,
```
Why does this take `bytes` as an argument?

stevvooe commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,497 @@
+		_, err := fw.driver.driveService.Files.Update(file.Id, dstFile).
+			Media(strings.NewReader(string(fw.bytes))).Do()
+		if err != nil {
+			fmt.Println("Close:unable to update")
```
Need to use logger here. Also, report the error.

stevvooe commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,497 @@
+
+// Delete recursively deletes all objects stored at "path" and its subpaths.
+func (d *driver) Delete(ctx context.Context, subPath string) error {
+	r, _ := d.driveService.Files.List().
```
Don't ignore the error.

stevvooe commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,497 @@
+	_, err := fw.driver.driveService.Files.Create(newFile).
+		Media(bytes.NewReader(fw.bytes)).Do()
+	if err != nil {
+		fmt.Println("Commit:Create file in drive failed", err)
```
Use a log message here. `log.WithError(err).Error(...)`

stevvooe commented on registry/storage/driver/gdrive/gdrive.go:
```diff
@@ -0,0 +1,497 @@
+		return 0, fmt.Errorf("already cancelled")
+	}
+
+	fw.bytes = append(fw.bytes, p...)
```
This is just buffering everything into memory. This should write the data through to the backend.

stevvooe commented:
@techknowlogick @pmjohann Thanks for the info! We'll see what can be done here.

I've put some review inline that need to be worked out before merging.

Who would maintain this in the future? In general, "the maintainers" won't have time for reports and fix for individual drivers, as they may not have the required environment. Someone will have to be available to handle bug reports and other problems.

pmjohann commented:
@stevvooe I would be more than happy to maintain it, but unfortunately I'm only onboarding the Go language itself. I think the learning curve won't be to steep, I have strong PHP and Node.js background.

Even until then, if I saw correctly, the PR author is @manoj535 . Could you take the responsibility? I'm sure a lot of students would appreciate it! Thank you!

manoj535 commented:
@pmjohann  , @stevvooe   We (me and @bharaththiruveedula) would like to support this driver in future too. We would like to maintain this driver in future to handle bug reports and other problems.

stevvooe commented:
@manoj535 Looks like there are a few build errors.

dmp42 commented:
@manoj535 are you still willing to work on this?

If so, can you rebase?

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "master" git@github.com:manoj535/distribution.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354132976
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

milosgajdos commented:
I dont think we'll consider adding more storage drivers to registry at this point, so I'm closing this PR. feel free to reopen if you disagree.
